File: scripts/generate-promo-images.js
#!/usr/bin/env node

/**
 * Promotional Images Generator using Puppeteer
 * Generates Chrome Web Store promo tiles (PNG) with modern styling
 * and a realistic "typing preview" editor mock.
 *
 * Requirements:
 *   npm i puppeteer
 *
 * Usage examples:
 *   node scripts/generate-promo-images.js
 *   node scripts/generate-promo-images.js --name="Demo Typer" --tagline="Human-like typing for demos" --theme=indigo
 *   OUTPUT_DIR=promo-images node scripts/generate-promo-images.js --emoji="⌨️"
 *
 * Options (CLI flags or env vars):
 *   --name / EXT_NAME           Extension name (default: "Demo Typer")
 *   --tagline / EXT_TAGLINE     Subtitle/tagline (default: "Realistic Typing Simulation")
 *   --badge / EXT_BADGE         Small badge label (default: "Chrome Extension")
 *   --emoji / EXT_EMOJI         Leading emoji (default: "⌨️")
 *   --theme / EXT_THEME         Color theme: indigo | teal | amber | purple (default: "indigo")
 *   --out   / OUTPUT_DIR        Output directory (default: "promo-images")
 */

const puppeteer = require('puppeteer');
const path = require('path');
const fs = require('fs');

// ---------- Config ----------

const argv = parseArgs(process.argv.slice(2));

const EXT_NAME    = argv.name    || process.env.EXT_NAME     || 'Demo Typer';
const EXT_TAGLINE = argv.tagline || process.env.EXT_TAGLINE  || 'Realistic Typing Simulation';
const EXT_BADGE   = argv.badge   || process.env.EXT_BADGE    || 'Chrome Extension';
const EXT_EMOJI   = argv.emoji   || process.env.EXT_EMOJI    || '⌨️';
const THEME_KEY   = (argv.theme  || process.env.EXT_THEME    || 'indigo').toLowerCase();
const OUTPUT_DIR  = argv.out     || process.env.OUTPUT_DIR   || 'promo-images';

const EXTENSION_PATH = path.join(__dirname, '..');

// Promo tile sizes required by Chrome Web Store
const PROMO_SIZES = {
  small:   { width: 440,  height: 280,  name: 'small_promo_tile_440x280.png' },
  marquee: { width: 1400, height: 560,  name: 'marquee_promo_tile_1400x560.png' }
};

// Minimal palettes with subtle gradients
const THEMES = {
  indigo: {
    grad: ['#667eea', '#764ba2'],
    glass: 'rgba(255,255,255,0.25)',
    glassBorder: 'rgba(255,255,255,0.35)',
    editorBg: '#0b1020',
    accent: '#a5b4fc'
  },
  teal: {
    grad: ['#06b6d4', '#3b82f6'],
    glass: 'rgba(255,255,255,0.23)',
    glassBorder: 'rgba(255,255,255,0.32)',
    editorBg: '#081a1a',
    accent: '#99f6e4'
  },
  amber: {
    grad: ['#f59e0b', '#ef4444'],
    glass: 'rgba(255,255,255,0.22)',
    glassBorder: 'rgba(255,255,255,0.30)',
    editorBg: '#1b1306',
    accent: '#fde68a'
  },
  purple: {
    grad: ['#8b5cf6', '#ec4899'],
    glass: 'rgba(255,255,255,0.25)',
    glassBorder: 'rgba(255,255,255,0.35)',
    editorBg: '#1a0e1e',
    accent: '#f5d0fe'
  }
};

const THEME = THEMES[THEME_KEY] || THEMES.indigo;

// ---------- Utils ----------

function parseArgs(args) {
  const out = {};
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a.startsWith('--')) {
      const [k, vRaw] = a.slice(2).split('=');
      if (vRaw !== undefined) {
        out[k] = vRaw;
      } else if (i + 1 < args.length && !args[i + 1].startsWith('--')) {
        out[k] = args[++i];
      } else {
        out[k] = true;
      }
    }
  }
  return out;
}

// Ensure output directory exists
function ensureDirectoryExists(dir) {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
}

// ---------- HTML Builders ----------

/**
 * Small Promo (440x280)
 * - Centered layout with tags
 * - Subtle glassy chips
 */
function getSmallPromoTileHTML() {
  const grad = `linear-gradient(135deg, ${THEME.grad[0]} 0%, ${THEME.grad[1]} 100%)`;
  return `
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8" />
<title>${safe(EXT_NAME)} - Small Promo</title>
<style>
  * { margin:0; padding:0; box-sizing:border-box; }
  html, body { width:440px; height:280px; }
  body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, sans-serif;
    background: ${grad};
    display:flex; align-items:center; justify-content:center; position:relative; overflow:hidden;
  }
  .bg-circle {
    position:absolute; border-radius:50%; background: rgba(255,255,255,0.08); filter: blur(0.2px);
  }
  .c1{ width:160px; height:160px; top:-40px; right:-30px; }
  .c2{ width:120px; height:120px; bottom:-30px; left:-20px; }
  .c3{ width:100px; height:100px; top:50%; right:10px; opacity:0.06; transform:translateY(-50%); }
  .container { text-align:center; z-index:1; padding:26px; }
  .icon { font-size:64px; margin-bottom:10px; text-shadow:0 4px 12px rgba(0,0,0,0.25); }
  h1 {
    color:#fff; font-size:34px; line-height:1.05; font-weight:800; letter-spacing:0.2px;
    text-shadow:0 2px 8px rgba(0,0,0,0.25); margin-bottom:6px;
  }
  .subtitle { color:rgba(255,255,255,0.96); font-size:15px; font-weight:600; margin-bottom:14px; }
  .tags { display:flex; gap:8px; justify-content:center; flex-wrap:wrap; }
  .tag {
    background:${THEME.glass}; color:#fff; padding:6px 12px; border-radius:14px; font-size:12px; font-weight:700;
    border:1px solid ${THEME.glassBorder}; backdrop-filter: blur(10px);
    box-shadow:0 2px 8px rgba(0,0,0,0.18);
  }
</style>
</head>
<body>
  <div class="bg-circle c1"></div>
  <div class="bg-circle c2"></div>
  <div class="bg-circle c3"></div>

  <div class="container">
    <div class="icon">${safe(EXT_EMOJI)}</div>
    <h1>${safe(EXT_NAME)}</h1>
    <p class="subtitle">${safe(EXT_TAGLINE)}</p>
    <div class="tags">
      <span class="tag">⚡ Adjustable speed</span>
      <span class="tag">🎯 Human-like</span>
      <span class="tag">⌨️ Shortcut-ready</span>
    </div>
  </div>
</body>
</html>`;
}

/**
 * Marquee Promo (1400x560)
 * - Left: circular emblem
 * - Right: headline, badge, features
 * - NEW: Editor mock with blinking caret + "typed" sample
 */
function getMarqueePromoTileHTML() {
  const grad = `linear-gradient(120deg, ${THEME.grad[0]} 0%, ${THEME.grad[1]} 100%)`;
  return `
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8" />
<title>${safe(EXT_NAME)} - Marquee Promo</title>
<style>
  * { margin:0; padding:0; box-sizing:border-box; }
  html, body { width:1400px; height:560px; }
  body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, sans-serif;
    background: ${grad};
    display:flex; align-items:center; position:relative; overflow:hidden;
  }
  .bg-shape { position:absolute; border-radius:50%; background:rgba(255,255,255,0.08); }
  .s1{ width:300px; height:300px; top:-100px; right:200px; }
  .s2{ width:240px; height:240px; bottom:-80px; left:150px; }
  .s3{ width:180px; height:180px; top:60px; right:50px; opacity:0.5; }
  .s4{ width:150px; height:150px; bottom:80px; right:300px; opacity:0.4; }
  .content { display:flex; align-items:center; width:100%; padding:0 80px; gap:60px; z-index:1; }
  .left { flex:0 0 420px; display:flex; flex-direction:column; align-items:center; gap:24px; }
  .icon-wrap { position:relative; display:inline-block; }
  .icon-bg {
    width:200px; height:200px; background:#fff; border-radius:50%;
    display:flex; align-items:center; justify-content:center; box-shadow:0 24px 70px rgba(0,0,0,0.35);
    position:relative;
  }
  .ring { position:absolute; border-radius:50%; border:2px solid rgba(255,255,255,0.35); animation:pulse 2.4s ease-in-out infinite; }
  .ring.r1{ width:236px; height:236px; }
  .ring.r2{ width:268px; height:268px; animation-delay:.35s; }
  @keyframes pulse { 0%,100%{ transform:scale(1); opacity:.55 } 50%{ transform:scale(1.05); opacity:.32 } }
  .bigicon { font-size:96px; }

  .right { flex:1; color:#fff; }
  .badge {
    display:inline-block; background:${THEME.glass}; color:#fff; padding:9px 22px; border-radius:22px;
    font-size:14px; font-weight:800; text-transform:uppercase; letter-spacing:.55px; margin-bottom:16px;
    border:1px solid ${THEME.glassBorder}; backdrop-filter: blur(10px); box-shadow:0 4px 12px rgba(0,0,0,.15);
  }
  h1 { font-size:76px; font-weight:900; margin-bottom:12px; text-shadow:0 4px 16px rgba(0,0,0,0.35); line-height:1.04; }
  .subtitle { font-size:26px; font-weight:500; margin-bottom:28px; opacity:.98; text-shadow:0 2px 8px rgba(0,0,0,.2); }

  /* Feature grid */
  .grid { display:grid; grid-template-columns:repeat(3,1fr); gap:18px; margin-bottom:26px; }
  .card {
    background: rgba(255,255,255,0.16); backdrop-filter: blur(10px);
    padding:18px; border-radius:14px; border:1px solid rgba(255,255,255,0.22);
    box-shadow:0 6px 18px rgba(0,0,0,0.12);
  }
  .fi { font-size:32px; display:block; margin-bottom:8px; }
  .ft { font-size:16px; font-weight:800; margin-bottom:4px; }
  .fd { font-size:12px; opacity:.92; line-height:1.4; }

  /* Editor mock with blinking caret + typed text */
  .editor {
    width:100%; max-width:760px; height:180px; border-radius:16px; overflow:hidden;
    box-shadow:0 18px 45px rgba(0,0,0,.35); border:1px solid rgba(255,255,255,.18);
    background: linear-gradient(180deg, ${THEME.editorBg} 0%, rgba(0,0,0,.75) 100%);
  }
  .ed-header {
    height:36px; background: rgba(255,255,255,0.06); border-bottom:1px solid rgba(255,255,255,0.12);
    display:flex; align-items:center; gap:8px; padding:0 12px; color:#cbd5e1; font-size:12px;
  }
  .dot { width:10px; height:10px; border-radius:50%; opacity:.9; }
  .dot.r{ background:#ef4444; } .dot.y{ background:#f59e0b; } .dot.g{ background:#10b981; }
  .ed-body {
    font-family: ui-monospace, SFMono-Regular, Menlo, Monaco, "Courier New", monospace;
    color:#e5e7eb; font-size:14px; line-height:1.6; padding:14px 16px;
    white-space:pre; position:relative;
  }
  .caret {
    display:inline-block; width:8px; height:1.2em; background:${THEME.accent}; margin-left:2px;
    animation:blink 1.1s step-end infinite; vertical-align:bottom;
  }
  @keyframes blink { 50% { opacity:0; } }
  .kw { color:#93c5fd; } .str{ color:#fca5a5; } .cm{ color:#9ca3af; font-style:italic; }
  .ac { color:${THEME.accent}; }
</style>
</head>
<body>
  <div class="bg-shape s1"></div>
  <div class="bg-shape s2"></div>
  <div class="bg-shape s3"></div>
  <div class="bg-shape s4"></div>

  <div class="content">
    <div class="left">
      <div class="icon-wrap">
        <div class="icon-bg">
          <div class="bigicon">${safe(EXT_EMOJI)}</div>
          <div class="ring r1"></div>
          <div class="ring r2"></div>
        </div>
      </div>
    </div>

    <div class="right">
      <div class="badge">🚀 ${safe(EXT_BADGE)}</div>
      <h1>${safe(EXT_NAME)}</h1>
      <p class="subtitle">${safe(EXT_TAGLINE)}</p>

      <div class="grid">
        <div class="card"><span class="fi">⚡</span><div class="ft">Adjustable Speed</div><div class="fd">Control typing pace in real time.</div></div>
        <div class="card"><span class="fi">🎯</span><div class="ft">Natural Patterns</div><div class="fd">Human-like pauses and bursts.</div></div>
        <div class="card"><span class="fi">⌨️</span><div class="ft">Easy Shortcuts</div><div class="fd">One-tap start and stop.</div></div>
      </div>

      <div class="editor">
        <div class="ed-header">
          <div class="dot r"></div><div class="dot y"></div><div class="dot g"></div>
          <span style="margin-left:8px; opacity:.9;">demo-typing.js</span>
        </div>
        <div class="ed-body">
<span class="cm">// Simulate realistic typing</span>
<span class="kw">const</span> typer <span class="kw">=</span> <span class="kw">new</span> <span class="ac">DemoTyper</span>(<span class="str">'Hello, world!'</span>, { speed: <span class="ac">120</span> });${' '}
<span class="kw">await</span> typer.<span class="ac">start</span>()<span class="kw">;</span> <span class="caret"></span>
        </div>
      </div>
    </div>
  </div>
</body>
</html>`;
}

// Basic HTML escaping for dynamic text
function safe(s) {
  return String(s)
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&#39;');
}

// ---------- Screenshot Helper ----------

async function capturePromoTile(browser, html, width, height, filename) {
  const page = await browser.newPage();
 
   // Set viewport to exact required dimensions
   await page.setViewport({ width, height, deviceScaleFactor: 1 });

  // Reduce animation variability
  await page.emulateMediaFeatures([
    { name: 'prefers-reduced-motion', value: 'reduce' }
  ]);

   await page.setContent(html, { waitUntil: 'networkidle0' });
 
   // Let fonts & animations settle briefly
   await new Promise(resolve => setTimeout(resolve, 600));

  const outputDir = path.join(EXTENSION_PATH, OUTPUT_DIR);
  const filepath = path.join(outputDir, filename);

  await page.screenshot({
    path: filepath,
    type: 'png',
    fullPage: false
  });

  console.log(`  ✓ Generated ${filename}`);
  await page.close();
}

// ---------- Main ----------

async function main() {
  console.log('🎨 Generating Chrome Web Store promotional images...\n');

  const outputDir = path.join(EXTENSION_PATH, OUTPUT_DIR);
  ensureDirectoryExists(outputDir);

  let browser;
  try {
    browser = await puppeteer.launch({
      headless: 'new',
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    });

    console.log('📸 Generating promotional tiles:');

    await capturePromoTile(
      browser,
      getSmallPromoTileHTML(),
      PROMO_SIZES.small.width,
      PROMO_SIZES.small.height,
      PROMO_SIZES.small.name
    );

    await capturePromoTile(
      browser,
      getMarqueePromoTileHTML(),
      PROMO_SIZES.marquee.width,
      PROMO_SIZES.marquee.height,
      PROMO_SIZES.marquee.name
    );

    console.log('\n✨ All promotional images generated successfully!');
    console.log(`📁 Images saved to: ${outputDir}`);
    console.log('\n📋 Chrome Web Store Requirements checked:');
    console.log('   • Small promo tile: 440×280 ✓');
    console.log('   • Marquee promo tile: 1400×560 ✓');
    console.log('   • Format: PNG ✓');

  } catch (error) {
    console.error('❌ Error generating promotional images:', error);
    process.exit(1);
  } finally {
    if (browser) {
      await browser.close();
    }
  }
}

// Run the script
if (require.main === module) {
  main();
}

module.exports = { main };

File: scripts/generate-promo-images.py
#!/usr/bin/env python3

"""
Chrome Web Store Promotional Images Generator
Generates screenshots and promotional tiles for the Demo Typer Extension

Requirements:
    pip install pillow

Usage:
    python3 scripts/generate-promo-images.py
    or
    ./scripts/generate-promo-images.py
"""

import os
from PIL import Image, ImageDraw, ImageFont

# Output directory
OUTPUT_DIR = 'promo-images'

# Color scheme
COLORS = {
    'primary': '#4A90E2',      # Blue
    'secondary': '#357ABD',    # Darker blue
    'accent': '#5BA3F5',       # Lighter blue
    'text': '#FFFFFF',         # White
    'text_dark': '#2C3E50',    # Dark gray
    'background': '#F8F9FA',   # Light gray
    'success': '#27AE60',      # Green
}

# Image configurations
SCREENSHOT_SIZES = [
    (1280, 800),
    (640, 400)
]

PROMO_SIZES = {
    'small': (440, 280),
    'marquee': (1400, 560)
}

def get_font(size, bold=False):
    """
    Get a system font with fallback to default
    
    Args:
        size (int): Font size
        bold (bool): Whether to use bold font
    
    Returns:
        PIL.ImageFont: Font object
    """
    font_paths = [
        # macOS
        '/System/Library/Fonts/Helvetica.ttc',
        '/System/Library/Fonts/SFNSDisplay.ttf',
        '/System/Library/Fonts/Supplemental/Arial.ttf',
        # Linux
        '/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf' if bold else '/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf',
        '/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf' if bold else '/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf',
        # Windows
        'C:\\Windows\\Fonts\\arialbd.ttf' if bold else 'C:\\Windows\\Fonts\\arial.ttf',
    ]
    
    for font_path in font_paths:
        if os.path.exists(font_path):
            try:
                return ImageFont.truetype(font_path, size)
            except:
                continue
    
    # Fallback to default
    return ImageFont.load_default()

def draw_text_centered(draw, text, position, font, fill, max_width=None):
    """
    Draw text centered at position
    
    Args:
        draw: ImageDraw object
        text (str): Text to draw
        position (tuple): (x, y) center position
        font: Font object
        fill: Text color
        max_width (int): Maximum width for text wrapping
    """
    if max_width:
        # Simple word wrap
        words = text.split()
        lines = []
        current_line = []
        
        for word in words:
            test_line = ' '.join(current_line + [word])
            bbox = draw.textbbox((0, 0), test_line, font=font)
            if bbox[2] - bbox[0] <= max_width:
                current_line.append(word)
            else:
                if current_line:
                    lines.append(' '.join(current_line))
                current_line = [word]
        if current_line:
            lines.append(' '.join(current_line))
        
        # Draw each line
        total_height = len(lines) * (font.size * 1.2)
        y = position[1] - total_height / 2
        
        for line in lines:
            bbox = draw.textbbox((0, 0), line, font=font)
            text_width = bbox[2] - bbox[0]
            x = position[0] - text_width / 2
            draw.text((x, y), line, fill=fill, font=font)
            y += font.size * 1.2
    else:
        bbox = draw.textbbox((0, 0), text, font=font)
        text_width = bbox[2] - bbox[0]
        text_height = bbox[3] - bbox[1]
        x = position[0] - text_width / 2
        y = position[1] - text_height / 2
        draw.text((x, y), text, fill=fill, font=font)

def generate_screenshot_1(size):
    """Generate screenshot showing typing simulation in action"""
    width, height = size
    img = Image.new('RGB', size, COLORS['background'])
    draw = ImageDraw.Draw(img)
    
    # Title
    title_font = get_font(int(height * 0.08), bold=True)
    draw_text_centered(draw, "Demo Typer Extension", (width // 2, int(height * 0.12)), title_font, COLORS['text_dark'])
    
    # Browser-like window
    window_padding = int(width * 0.1)
    window_y = int(height * 0.22)
    window_height = int(height * 0.65)
    
    # Window background
    draw.rectangle([window_padding, window_y, width - window_padding, window_y + window_height], 
                   fill='#FFFFFF', outline=COLORS['secondary'], width=2)
    
    # Browser address bar
    bar_height = int(height * 0.05)
    draw.rectangle([window_padding, window_y, width - window_padding, window_y + bar_height],
                   fill=COLORS['background'], outline=COLORS['secondary'], width=1)
    
    # Address bar text
    addr_font = get_font(int(height * 0.025))
    draw.text((window_padding + 10, window_y + bar_height // 4), "https://example.com", 
              fill=COLORS['text_dark'], font=addr_font)
    
    # Content area with typing effect
    content_y = window_y + bar_height + 30
    content_font = get_font(int(height * 0.04))
    
    typing_text = "Hello World!"
    cursor_x = window_padding + 30
    
    for i, char in enumerate(typing_text):
        color = COLORS['success'] if i < len(typing_text) - 3 else COLORS['text_dark']
        draw.text((cursor_x, content_y), char, fill=color, font=content_font)
        bbox = draw.textbbox((0, 0), char, font=content_font)
        cursor_x += (bbox[2] - bbox[0])
    
    # Blinking cursor
    draw.rectangle([cursor_x + 2, content_y, cursor_x + 6, content_y + int(height * 0.04)],
                   fill=COLORS['primary'])
    
    # Description
    desc_font = get_font(int(height * 0.03))
    draw_text_centered(draw, "Realistic typing simulation for impressive demos", 
                      (width // 2, int(height * 0.95)), desc_font, COLORS['text_dark'])
    
    return img

def generate_screenshot_2(size):
    """Generate screenshot showing extension popup/settings"""
    width, height = size
    img = Image.new('RGB', size, COLORS['background'])
    draw = ImageDraw.Draw(img)
    
    # Title
    title_font = get_font(int(height * 0.08), bold=True)
    draw_text_centered(draw, "Easy Configuration", (width // 2, int(height * 0.12)), title_font, COLORS['text_dark'])
    
    # Settings panel
    panel_width = int(width * 0.5)
    panel_height = int(height * 0.65)
    panel_x = (width - panel_width) // 2
    panel_y = int(height * 0.22)
    
    # Panel background
    draw.rounded_rectangle([panel_x, panel_y, panel_x + panel_width, panel_y + panel_height],
                          radius=10, fill='#FFFFFF', outline=COLORS['primary'], width=3)
    
    # Panel header
    header_font = get_font(int(height * 0.05), bold=True)
    draw.text((panel_x + 20, panel_y + 20), "⚙️ Settings", fill=COLORS['primary'], font=header_font)
    
    # Settings options
    option_font = get_font(int(height * 0.035))
    option_y = panel_y + 100
    option_spacing = int(height * 0.08)
    
    options = [
        "⌨️  Typing Speed: Adjustable",
        "⏱️  Delay: Customizable",
        "🎯  Target: Any Text Field",
        "🔄  Keyboard Shortcuts",
    ]
    
    for option in options:
        draw.text((panel_x + 40, option_y), option, fill=COLORS['text_dark'], font=option_font)
        option_y += option_spacing
    
    # Description
    desc_font = get_font(int(height * 0.03))
    draw_text_centered(draw, "Customize typing speed and behavior to match your demo style", 
                      (width // 2, int(height * 0.95)), desc_font, COLORS['text_dark'])
    
    return img

def generate_screenshot_3(size):
    """Generate screenshot showing use case"""
    width, height = size
    img = Image.new('RGB', size, COLORS['background'])
    draw = ImageDraw.Draw(img)
    
    # Title
    title_font = get_font(int(height * 0.08), bold=True)
    draw_text_centered(draw, "Perfect for Presentations", (width // 2, int(height * 0.12)), title_font, COLORS['text_dark'])
    
    # Three feature boxes
    box_width = int(width * 0.25)
    box_height = int(height * 0.5)
    box_y = int(height * 0.25)
    spacing = int(width * 0.04)
    start_x = (width - (3 * box_width + 2 * spacing)) // 2
    
    features = [
        ("💻", "Code Demos", "Perfect for live\ncoding sessions"),
        ("🎓", "Tutorials", "Create engaging\nvideo tutorials"),
        ("🎪", "Presentations", "Impress your\naudience"),
    ]
    
    icon_font = get_font(int(height * 0.12))
    title_feat_font = get_font(int(height * 0.045), bold=True)
    desc_feat_font = get_font(int(height * 0.03))
    
    for i, (icon, title, desc) in enumerate(features):
        box_x = start_x + i * (box_width + spacing)
        
        # Box
        draw.rounded_rectangle([box_x, box_y, box_x + box_width, box_y + box_height],
                              radius=10, fill='#FFFFFF', outline=COLORS['primary'], width=2)
        
        # Icon
        draw_text_centered(draw, icon, (box_x + box_width // 2, box_y + int(box_height * 0.25)),
                          icon_font, COLORS['primary'])
        
        # Title
        draw_text_centered(draw, title, (box_x + box_width // 2, box_y + int(box_height * 0.55)),
                          title_feat_font, COLORS['text_dark'])
        
        # Description
        desc_lines = desc.split('\n')
        desc_y = box_y + int(box_height * 0.7)
        for line in desc_lines:
            draw_text_centered(draw, line, (box_x + box_width // 2, desc_y),
                              desc_feat_font, COLORS['text_dark'])
            desc_y += int(height * 0.04)
    
    # Description
    desc_font = get_font(int(height * 0.03))
    draw_text_centered(draw, "Make your demonstrations look professional and polished", 
                      (width // 2, int(height * 0.92)), desc_font, COLORS['text_dark'])
    
    return img

def generate_small_promo_tile():
    """Generate small promotional tile (440x280)"""
    width, height = PROMO_SIZES['small']
    img = Image.new('RGB', (width, height), COLORS['primary'])
    draw = ImageDraw.Draw(img)
    
    # Large "T" icon
    icon_font = get_font(int(height * 0.4), bold=True)
    draw_text_centered(draw, "T", (width // 2, int(height * 0.35)), icon_font, COLORS['text'])
    
    # Title
    title_font = get_font(int(height * 0.12), bold=True)
    draw_text_centered(draw, "Demo Typer", (width // 2, int(height * 0.68)), title_font, COLORS['text'])
    
    # Subtitle
    subtitle_font = get_font(int(height * 0.07))
    draw_text_centered(draw, "Realistic Typing Simulation", (width // 2, int(height * 0.85)), 
                      subtitle_font, COLORS['text'])
    
    return img

def generate_marquee_promo_tile():
    """Generate marquee promotional tile (1400x560)"""
    width, height = PROMO_SIZES['marquee']
    img = Image.new('RGB', (width, height), COLORS['primary'])
    draw = ImageDraw.Draw(img)
    
    # Split into left and right sections
    left_width = int(width * 0.4)
    
    # Left section - Icon and branding
    icon_font = get_font(int(height * 0.4), bold=True)
    draw_text_centered(draw, "T", (left_width // 2, height // 2), icon_font, COLORS['text'])
    
    # Right section - Text content
    right_x = left_width + int(width * 0.05)
    
    # Title
    title_font = get_font(int(height * 0.12), bold=True)
    draw.text((right_x, int(height * 0.2)), "Demo Typer Extension", fill=COLORS['text'], font=title_font)
    
    # Subtitle
    subtitle_font = get_font(int(height * 0.06))
    draw.text((right_x, int(height * 0.4)), "Create impressive demos with realistic typing simulation", 
              fill=COLORS['text'], font=subtitle_font)
    
    # Features
    feature_font = get_font(int(height * 0.045))
    feature_y = int(height * 0.58)
    features = [
        "✓ Adjustable typing speed",
        "✓ Natural typing patterns",
        "✓ Easy keyboard shortcuts",
    ]
    
    for feature in features:
        draw.text((right_x + 20, feature_y), feature, fill=COLORS['text'], font=feature_font)
        feature_y += int(height * 0.1)
    
    return img

def convert_to_24bit_png(img):
    """
    Convert image to 24-bit PNG (RGB without alpha)
    
    Args:
        img: PIL Image
    
    Returns:
        PIL.Image: Image in RGB mode
    """
    if img.mode != 'RGB':
        return img.convert('RGB')
    return img

def main():
    """Main function to generate all promotional images"""
    # Get the output directory path
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output_dir = os.path.join(os.path.dirname(script_dir), OUTPUT_DIR)
    
    # Ensure output directory exists
    os.makedirs(output_dir, exist_ok=True)
    
    print('🎨 Generating Chrome Web Store promotional images...\n')
    
    # Generate screenshots
    print('📸 Generating screenshots:')
    
    screenshot_generators = [
        generate_screenshot_1,
        generate_screenshot_2,
        generate_screenshot_3,
    ]
    
    for i, generator in enumerate(screenshot_generators, 1):
        # Generate both sizes
        for j, size in enumerate(SCREENSHOT_SIZES, 1):
            try:
                img = generator(size)
                img = convert_to_24bit_png(img)
                filename = f'screenshot_{i}_{size[0]}x{size[1]}.png'
                filepath = os.path.join(output_dir, filename)
                img.save(filepath, 'PNG')
                print(f'  ✓ {filename}')
            except Exception as e:
                print(f'  ✗ Failed to generate {filename}: {str(e)}')
    
    # Generate promotional tiles
    print('\n🎯 Generating promotional tiles:')
    
    try:
        # Small promo tile
        small_tile = generate_small_promo_tile()
        small_tile = convert_to_24bit_png(small_tile)
        small_path = os.path.join(output_dir, 'small_promo_tile_440x280.png')
        small_tile.save(small_path, 'PNG')
        print(f'  ✓ small_promo_tile_440x280.png')
    except Exception as e:
        print(f'  ✗ Failed to generate small promo tile: {str(e)}')
    
    try:
        # Marquee promo tile
        marquee_tile = generate_marquee_promo_tile()
        marquee_tile = convert_to_24bit_png(marquee_tile)
        marquee_path = os.path.join(output_dir, 'marquee_promo_tile_1400x560.png')
        marquee_tile.save(marquee_path, 'PNG')
        print(f'  ✓ marquee_promo_tile_1400x560.png')
    except Exception as e:
        print(f'  ✗ Failed to generate marquee promo tile: {str(e)}')
    
    print(f'\n✨ All promotional images generated successfully!')
    print(f'📁 Images saved to: {output_dir}')
    print(f'\n📋 Chrome Web Store Requirements:')
    print(f'   • Screenshots: 1280x800 or 640x400 (up to 5 max)')
    print(f'   • Small promo tile: 440x280')
    print(f'   • Marquee promo tile: 1400x560')
    print(f'   • All images are 24-bit PNG (no alpha)')

if __name__ == '__main__':
    main()

